File: src/test/java/neoflix/_07_FavoritesListTest.java
// Task: Rewrite the AuthService to allow users to authenticate against the database
// Outcome: A user will be able to authenticate against their database record
package neoflix;

import neoflix.services.FavoriteService;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.neo4j.driver.Driver;
import org.neo4j.driver.Values;

import static org.junit.jupiter.api.Assertions.*;

class _07_FavoritesListTest {
    private static Driver driver;

    private static final String toyStory = "862";
    private static final String goodfellas = "769";
    private static final String userId = "9f965bf6-7e32-4afb-893f-756f502b2c2a";
    private static final String email = "graphacademy.favorite@neo4j.com";

    @BeforeAll
    static void initDriver() {
        AppUtils.loadProperties();
        driver = AppUtils.initDriver();

        try (var session = driver.session()) {
            session.writeTransaction(tx -> tx.run("""
                    MERGE (u:User {userId: $userId}) SET u.email = $email
                    """, Values.parameters("userId", userId, "email", email)));
        }
    }

    @AfterAll
    static void closeDriver() {
        driver.close();
    }

    @Test
    void notFoundIfMovieOrUserNotExist() {
        FavoriteService favoriteService = new FavoriteService(driver);

        try {
            favoriteService.add("unknown", "x999");
            fail("Adding favorite with unknown userId or movieId should fail");
        } catch (Exception e) {
            assertEquals("Couldn't create a favorite relationship for User unknown and Movie x999", e.getMessage());
        }
    }

    @BeforeEach
    void removeFavorites() {
        try (var session = driver.session()) {
            session.writeTransaction(tx ->
                    tx.run("MATCH (u:User {userId: $userId})-[r:HAS_FAVORITE]->(m:Movie) DELETE r",
                            Values.parameters("userId", userId)));
        }
    }

    @Test
    void saveMovieToUserFavorites() {
        FavoriteService favoriteService = new FavoriteService(driver);

        var output = favoriteService.add(userId, toyStory);

        assertNotNull(output);
        assertEquals(toyStory, output.get("tmdbId"));
        assertTrue((Boolean)output.get("favorite"), "toy story is favorite");

        var favorites = favoriteService.all(userId, new Params(null, Params.Sort.title, Params.Order.DESC, 10, 0));

        var movieFavorite = favorites.stream().allMatch(movie -> movie.get("tmdbId").equals(toyStory));
        assertTrue(movieFavorite, "only toy story is favorite movie");
    }

    @Test
    void addAndRemoveMovieFromFavorites() {
        FavoriteService favoriteService = new FavoriteService(driver);

        var add = favoriteService.add(userId, goodfellas);
        assertEquals(goodfellas, add.get("tmdbId"));
        assertTrue((Boolean)add.get("favorite"), "goodfellas is favorite");

        var addCheck = favoriteService.all(userId, new Params(null, Params.Sort.title, Params.Order.DESC, 10, 0));
        var found = addCheck.stream().allMatch(movie -> movie.get("tmdbId").equals(goodfellas));
        assertTrue(found, "onyl goodfellas is favorite");

        var remove = favoriteService.remove(userId, goodfellas);
        assertEquals(goodfellas, remove.get("tmdbId"));
        assertEquals(false, remove.get("favorite"), "goodfellas is not a favorite anymore");

        var removeCheck = favoriteService.all(userId, new Params(null, Params.Sort.title, Params.Order.DESC, 10, 0));
        assertTrue(removeCheck.isEmpty(), "no favorite movies anymore");
    }
}

File: pkg/challenges/07_favorites_list_test.go
package challenges_test

import (
	"testing"

	"github.com/neo4j-graphacademy/neoflix/pkg/fixtures"

	"github.com/neo4j-graphacademy/neoflix/pkg/config"
	"github.com/neo4j-graphacademy/neoflix/pkg/routes/paging"
	"github.com/neo4j-graphacademy/neoflix/pkg/services"
	"github.com/neo4j/neo4j-go-driver/v4/neo4j"
)

func TestMyFavoritesList(t *testing.T) {
	// Load Settings
	settings, err := config.ReadConfig("../../config.json")
	assertNilError(t, err)

	// Init Driver
	driver, err := config.NewDriver(settings)
	assertNilError(t, err)

	defer func() {
		assertNilError(t, driver.Close())
	}()

	// Create Services
	service := services.NewFavoriteService(
		&fixtures.FixtureLoader{Prefix: "../.."},
		driver)

	assertNotNil(t, service)

	toyStory := "862"
	goodfellas := "769"
	userId := "9f965bf6-7e32-4afb-893f-756f502b2c2a"
	email := "graphacademy.favorite@neo4j.com"

	// Create the User
	session := driver.NewSession(neo4j.SessionConfig{})
	session.Run(`
		MERGE (u:User {userId: $userId}) SET u.email = $email
		FOREACH (r IN [(u)-[r:HAS_FAVORITE]->() | r ] | DELETE r)
	`, map[string]interface{}{"userId": userId, "email": email})

	// Should throw an error if user or movie do not exist
	unknown, err := service.Save("unknown", "x999")
	assertFalse(t, unknown != nil)
	assertNotNil(t, err)

	unknownMovie, err := service.Save(userId, "x999")
	assertFalse(t, unknownMovie != nil)
	assertNotNil(t, err)

	unknownUser, err := service.Save("unknown", toyStory)
	assertFalse(t, unknownUser != nil)
	assertNotNil(t, err)

	// Add to list
	saved, err := service.Save(userId, toyStory)
	assertNilError(t, err)
	assertEquals(t, toyStory, saved["tmdbId"])
	assertEquals(t, true, saved["favorite"])

	all, err := service.FindAllByUserId(userId, paging.NewPaging("", "createdAt", "desc", 0, 1))

	assertNilError(t, err)
	assertEquals(t, len(all), 1)
	assertEquals(t, all[0]["tmdbId"], toyStory)

	remove, err := service.Delete(userId, toyStory)
	assertNilError(t, err)
	assertEquals(t, toyStory, remove["tmdbId"])
	assertEquals(t, false, remove["favorite"])

	// Add & Remove from list
	add, err := service.Save(userId, goodfellas)

	assertNilError(t, err)
	assertEquals(t, goodfellas, add["tmdbId"])
	assertEquals(t, true, add["favorite"])

	removeGoodfellas, err := service.Delete(userId, goodfellas)
	assertNilError(t, err)
	assertEquals(t, goodfellas, removeGoodfellas["tmdbId"])
	assertEquals(t, false, removeGoodfellas["favorite"])

	// Re-add the Toy Story Favorite for test
	readd, err := service.Save(userId, toyStory)
	assertNilError(t, err)
	assertNotNil(t, readd)
}
